lib/services/network_dio.dart

```
@@ -0,0 +1,58 @@
+import 'package:dio/dio.dart';
+
+import '../config/config.dart';
+import '../config/config_keys.dart';
+
+
+class NetworksDio {
+  static const String urlDomain =
+      'https://api.openweathermap.org/data/2.5/weather?';
+
+  Dio dio = Dio();
+
+  String byCity(String city) {
+    return '${urlDomain}q=$city&appid=${getApiKey()}&units=metric';
+  }
+
+  String getApiKey() {
+    final Config config = Config.manager;
+    return config.get(ConfigKeys.apiKey);
+  }
+
+  String byLatLong({required double latitude, required double longitude}) {
+    return '${urlDomain}lat=$latitude&lon=$longitude&appid=${getApiKey()}&units=metric';
+  }
+
+  Future<Map<String, dynamic>?> getDataByCity(String cityName) async {
+    try {
+      final response = await dio.get(byCity(cityName));
+      return decodeResponse(response);
+    } catch (e) {
+      print("Error: $e");
```

nstead of just printing the error message, you can throw a custom exception with a more descriptive error message. This will make it easier to debug the error and handle it in a more meaningful way.

```
Future<Map<String, dynamic>?> getDataByCity(String cityName) async {
    try {
      final response = await dio.get(byCity(cityName));
      return decodeResponse(response);
    } on DioError catch (e) {
      if (e.type == DioErrorType.connectTimeout ||
          e.type == DioErrorType.receiveTimeout ||
          e.type == DioErrorType.sendTimeout) {
        throw Exception('Connection timed out. Please try again later.');
      } else if (e.type == DioErrorType.response) {
        throw Exception('Server error. Please try again later.');
      } else {
        throw Exception('Unknown error occurred. Please try again later.');
      }
    } catch (e) {
      throw Exception('Unknown error occurred. Please try again later.');
    }
  }
```
and you can also display it in snackbar which helps in better user experience